### PR TITLE
KEYCLOAK-11491 Add index on event_entity's event_time column

### PR DESCRIPTION
**Pull request by knutz3n**

### CODE REVIEW

File: model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml]: Please change author to "keycloak" and id to "8.0.0".

**Review by stianst — CHANGES_REQUESTED** [review]
LGTM, except one minor thing

File: model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml
Comment (knutz3n) [reply model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml]: Updated the commit

**Review by stianst — DISMISSED** [review]

**Comment by stianst** [comment]
@mposolda @hmlnarik do we need to run db test matrix on this before merging or is it straightforward not to just go for it?

**Review by hmlnarik — CHANGES_REQUESTED** [review]
This should not get in unless impact on existing installations is considered. I'm rather against it for the same reasons as in https://github.com/keycloak/keycloak/pull/4394#issuecomment-332471832

**Comment by stianst** [comment]
@hmlnarik auch, that is rather complicated. We've added indexes in the past without considering that, but you're correct I'm afraid ;)

**Comment by knutz3n** [comment]
@hmlnarik Agree regarding the considerations related to the upgrade process. However, the current situation is that when the scheduled job to expire events is running, it causes insertion of new events to hang for several seconds. Do you have any preferences on how to improve this?

**Comment by knutz3n** [comment]
Is adding database-specific commands an acceptable solution? I think for most queries, indexing on event_time is enough. The resulting set after it is filtered by event_time is small enough that it would work.

**Comment by hmlnarik** [comment]
> Is adding database-specific commands an acceptable solution? I think for most queries, indexing on event_time is enough. The resulting set after it is filtered by event_time is small enough that it would work.

I agree that the issue needs solution. IMO the best what we could do is to introduce a [custom liquibase change](https://www.liquibase.org/documentation/changes/custom_change.html) which would be implemented for lazy indexing, and use that change in the changesets. Reason for that is possibility of using custom schema which is hard to use in liquibase <sql>. Then we could use it for other columns as well.

**Comment by knutz3n** [comment]
I have updated the PR to implement a generic LazyCreateIndexCustomChange class which can be used to create indexes without locking on Postgres, Oracle, MySQL and MsSQL databases. Is this something which can be tested in your db test matrix?

**Comment by knutz3n** [comment]
Postgres fails when the concurrent flag is set.  I will close this PR and reopen it later we I find a workaround.

File: model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java
Comment (thielef) [inline model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java]: Does this work for all MSSQL version? according to their documenation, ONLINE=ON is only supported for Enterprise MSSQL -- which some customers might not have.

**Review by thielef — COMMENTED** [review]
I think there is also a problem for MSSQL

File: model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java
Comment (knutz3n) [reply model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java]: Yes, I agree, only Enterprise MSSQL supports the online flag. I think the Keycloak core team  should reconsider the conclusion regarding these indexes. There are several issues related to missing indexes and several of them causes real problems. At least a constructive reply would be helpful. Right no it seems like the issues are just ignored.

Comment (hmlnarik) [reply model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java]: I agree that the missing indices are painful. This issue is not being ignored, however there are deployments that could have been disrupted by forced creation of the indices.

Clearly there is no simple automated way to create the indices due to differences among the databases. Perhaps we can do the following liquibase extension similar to what you have implemented:
* An index can just be created in the foreground if the affected table rows count is smaller than a given configurable number with a reasonable default value (let say 500.000 rows)
* Otherwise a warning with the SQL message is logged that contains a SQL command recommended to be applied manually by the DB admin.

Would that work for you, @thielef, @knutz3n, @gtudan ?

Comment (thielef) [reply model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java]: +1. Sounds like a good compromise. On top of that the official documentation (+release notes) should get a chapter about this fact with arguments why Keycloak cannot apply this automatically to the whole bunch of existing setups.

Comment (knutz3n) [reply model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java]: Yes, I think that is a fair compromise. I have a couple of questions, though.
1. Where should the configurable number be be configured? Standalone.xml?
2. Will Liquibase have knowledge of the index after it is applied manually? Does it need to?
3. Would it be an option to require the DB admin to add the index before upgrading? So Liquibase can just ensure that the index exists and if not, and there are less than x rows, it will apply the index?

Also, regarding Postgres which has support to add indexes in the background, it would be nice to get it to work with adding indexes concurrently. However, Postgres requires this to be done outside an transaction. When I tested it in Keycloak I got a deadlock even when the index was complete. Is this something we could look into? I will be happy to demonstrate it to you in a video meeting or something.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml
Comment (jdevreese) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-8.0.0.xml]: > I think for most queries, indexing on event_time is enough. The resulting set after it is filtered by event_time is small enough that it would work.

Please add the REALM_ID column to the index (and put it first).
It may not matter much if you only have a handful of realms, but I'm working on an environment with 600+ realms (which will increase further over the next few years to about a thousand).

**Comment by hSATAC** [comment]
Any progress on this issue? Is the current solution still create index manually?

**Pull request closed without merging**
